android/src/main/java/com/oblador/keychain/KeychainModule.java

```diff
@@ -37,99 +41,108 @@ public String getName() {
 
     public KeychainModule(ReactApplicationContext reactContext) {
         super(reactContext);
+        prefsStorage = new PrefsStorage(reactContext);
 
-        KeyChain keyChain = new SharedPrefsBackedKeyChain(getReactApplicationContext(), CryptoConfig.KEY_256);
-        crypto = AndroidConceal.get().createDefaultCrypto(keyChain);
-        prefs = this.getReactApplicationContext().getSharedPreferences(KEYCHAIN_DATA, Context.MODE_PRIVATE);
+        addCipherStorageToMap(new CipherStorageFacebookConceal(reactContext));
+        addCipherStorageToMap(new CipherStorageKeystoreAESCBC());
     }
 
-    @ReactMethod
-    public void setGenericPasswordForOptions(String service, String username, String password, Promise promise) {
-        if (!crypto.isAvailable()) {
-            Log.e(KEYCHAIN_MODULE, "Crypto is missing");
-            promise.reject("KeychainModule: crypto is missing");
-            return;
-        }
-        if (username == null || username.isEmpty() || password == null || password.isEmpty()) {
-            Log.e(KEYCHAIN_MODULE, "you passed empty or null username/password");
-            promise.reject("KeychainModule: you passed empty or null username/password");
-            return;
-        }
-        service = service == null ? EMPTY_STRING : service;
-        //Log.d("Crypto", service + username + password);
-
-        Entity userentity = Entity.create(KEYCHAIN_DATA + ":" + service + "user");
-        Entity pwentity = Entity.create(KEYCHAIN_DATA + ":" + service + "pass");
-
-
-        String encryptedUsername = encryptWithEntity(username, userentity, promise);
-        String encryptedPassword = encryptWithEntity(password, pwentity, promise);
-
-        SharedPreferences.Editor prefsEditor = prefs.edit();
-        prefsEditor.putString(service + ":u", encryptedUsername);
-        prefsEditor.putString(service + ":p", encryptedPassword);
-        prefsEditor.apply();
-        Log.d(KEYCHAIN_MODULE, "saved the data");
-        promise.resolve("KeychainModule saved the data");
+    private void addCipherStorageToMap(CipherStorage cipherStorage) {
+        cipherStorageMap.put(cipherStorage.getCipherStorageName(), cipherStorage);
     }
 
-    private String encryptWithEntity(String toEncypt, Entity entity, Promise promise) {
+    @ReactMethod
+    public void setGenericPasswordForOptions(String service, String username, String password, Promise promise) {
```

Comment:
You are right. I wanted a null `service` to be handled internally in each CipherStorage implementation, because it might differ from implementation to implementation - but of course the prefStorage should then also handle a null `service`.